### ISSUE TITLE
Nan values in confusion matrix

### ISSUE THREAD

**Issue by dervirvel** [voldemortX/DST-CBC#8]
Hello, 
i'm using a segmentation dataset with two classes and grayscale images. I'm duplicating the channels of the image with 
`elif pic.mode == 'L':
      img = torch.from_numpy(np.array(pic, np.uint8, copy=False)).expand([3, 636, 636]).reshape(-1)`
While training i get only accuracy values for one of the classes with nan values for the other:
`average row correct: ['99.52', 'nan']`. 

Do you have any idea what i'd might have done wrong/missed? Thanks in advance!

**Comment by voldemortX** [comment]
@dervirvel Hi! Is that training accuracy or test accuracy? Only cases when I get a NaN in confusion matrices are: when my model did not correctly classify a single annotated pixel from that class. That happens quite often in SSL, since the annotations are scarce already.
Although I've not used grayscale images before, but I think it is probably unrelated to this problem.

**Comment by voldemortX** [comment]
Also, is your dataset highly imbalanced between the 2 classes?

**Comment by dervirvel** [comment]
Hi, thanks for the fast reply! 
It's training accuracy. That might be one explanation, though it would be weird, since the dataset is balanced, but unfortunately extremely small (< 200)

**Comment by voldemortX** [comment]
Yes it is a bit weird, can you check if your labels are correct (0, 1, 255), and does this NaN value keep showing continuously in training?

**Comment by dervirvel** [comment]
Labels are correct (0,255), since everything is either foreground or background. It happens continuously. I'll check the predictions

**Comment by voldemortX** [comment]
> Labels are correct (0,255), since everything is either foreground or background. It happens continuously. I'll check the predictions

Just checking to be sure: in your dataset, label value 0 means foreground and 255 means background?

**Comment by dervirvel** [comment]
> > Labels are correct (0,255), since everything is either foreground or background. It happens continuously. I'll check the predictions
> 
> Just checking to be sure: in your dataset, label value 0 means foreground and 255 means background?

Yes, exactly.

**Comment by voldemortX** [comment]
@dervirvel In that case, I think the problem is found.
It seems you should use label value 1 for background, instead of 255. I set 255 as ignored pixels...

**Comment by dervirvel** [comment]
> @dervirvel In that case, I think the problem is found.
> It seems you should use label value 1 for background, instead of 255. I set 255 as ignored pixels...

Changed 255 -> 1, works now. Thanks very much! I remember reading the "#ignore 255" comment in commons.py, but didn't connect the dots. Is there a reason for ignoring 255?

**Comment by voldemortX** [comment]
> Changed 255 -> 1, works now. Thanks very much! I remember reading the "#ignore 255" comment in commons.py, but didn't connect the dots. Is there a reason for ignoring 255?

Well, for instance, when you rotate an image, some part of the image will have no label, thus a "ignore" is needed. For most multi-class tasks, labels are 01234..., 255 is the safest for "ignore" in uint8.
